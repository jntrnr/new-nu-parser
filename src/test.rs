use crate::resolver::Resolver;
use crate::{compiler::Compiler, parser::Parser};
use std::path::Path;

fn evaluate_example(fname: &Path) -> String {
    let mut compiler = Compiler::new();
    let contents = std::fs::read(fname).expect("We only run tests found by glob");

    // Homogenize line endings from CRLF to LF to make sure tests work both on Windows and Linux/macOS
    let contents = String::from_utf8(contents)
        .unwrap()
        .replace("\r\n", "\n")
        .as_bytes()
        .to_owned();

    let span_offset = compiler.span_offset();
    compiler.add_file(&fname.to_string_lossy(), &contents);

    let parser = Parser::new(compiler, span_offset);
    compiler = parser.parse();

    let mut result = compiler.display_state();

    if !compiler.errors.is_empty() {
        return result;
    }

    let mut resolver = Resolver::new(&compiler);
    resolver.resolve();
    result.push_str(&resolver.display_state());

    compiler.merge_name_bindings(resolver.to_name_bindings());

    result
}

#[test]
fn test_node_output() {
    insta::glob!("../tests", "*.nu", |path| {
        insta::assert_snapshot!(evaluate_example(path));
    });
}
