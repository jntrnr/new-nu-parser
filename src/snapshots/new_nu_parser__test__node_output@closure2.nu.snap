---
source: src/test.rs
expression: evaluate_example(path)
input_file: tests/closure2.nu
---
0: Variable (4 to 5)
1: Int (8 to 10)
2: Let { variable_name: NodeId(0), ty: None, initializer: NodeId(1), is_mutable: false } (0 to 10)
3: Variable (15 to 16)
4: Int (19 to 21)
5: Let { variable_name: NodeId(3), ty: None, initializer: NodeId(4), is_mutable: false } (11 to 21)
6: Variable (27 to 29)
7: Plus (30 to 31)
8: Variable (32 to 34)
9: BinaryOp { lhs: NodeId(6), op: NodeId(7), rhs: NodeId(8) } (27 to 34)
10: Block(BlockId(0)) (27 to 35)
11: Closure { params: None, block: NodeId(10) } (23 to 36)
12: Block(BlockId(1)) (0 to 37)

